categorize_adjectives/hierarchical.py

```diff
@@ -0,0 +1,92 @@
+"""Clustering the adjectives in a hierarchical manner"""
+import os
+import pickle
+
+def cluster_cluster_distance(cluster_1, cluster_2):
+    """Calculate distance between clusters
+
+    It is defined as average over distance between adj1, adj2
+    pairs where adj1 belongs to cluster_1 and adj2 belongs to cluster_2
+
+    returns distance between the two clusters
+    """
+    average_distance = 0
+    for adj_1 in cluster_1:
+        for adj_2 in cluster_2:
+            average_distance += distance_metric[(adj_1, adj_2)]
+    average_distance /= (len(cluster_1)*len(cluster_2))
+    return average_distance
+
+def minimize_clusters(clusters, min_threshold, max_threshold):
+    """Minimize the number of clusters by combining clusters that are at minimum distance
+
+    min_threshold : minimum threshold for combining of clusters
+    max_threshold: maximum threshold for combining of clusters
+
+    returns the list of minimized clusters
+    """
+    is_changing = True
+    while is_changing:
+        is_changing = False
+        for _, cluster1 in enumerate(clusters):
+            min_distance = 1
+            cluster_idx = -1
+            for idx1, cluster2 in enumerate(clusters):
+                if cluster2 == cluster1:
+                    continue
+                distance = cluster_cluster_distance(cluster1, cluster2)
+                if distance < min_distance:
+                    min_distance = distance
+                    cluster_idx = idx1
+            len1 = len(cluster1)
+            len2 = len(clusters[cluster_idx])
+            # We can also vary the condition for clustering here
+            if min_distance < min(min_threshold + 0.005*(len1*len2 - len1 -len2 + 1),
+                                  max_threshold):
+                is_changing = True
+                cluster2 = clusters[cluster_idx]
+                new_cluster = cluster1 | cluster2
+                clusters.remove(cluster1)
+                clusters.remove(cluster2)
+                clusters.append(new_cluster)
+                break
+            if is_changing:
+                break
+    return clusters
+
+def save_as_csv(clusters, file_name):
+    """Saves the clusters in hierarchial_clusters.csv file"""
+    file = open(file_name, "w")
+    for cluster in clusters:
+        sentence = ""
+        for adjective in cluster:
+            sentence += adjective + ","
+        sentence = sentence[:-1]
+        sentence += "\n"
+        file.write(sentence)
+    file.close()
+
+if __name__ == "__main__":
+    # Folder where generated files are stored
+    generated_file = os.path.join(os.getcwd(), '..', 'generated_files/')
+
+    with open(generated_file + "adj_distance.dat", 'rb') as f:
+        distance_metric = pickle.load(f)
+
+    adj_list = []
+    for adj1, adj2 in distance_metric.keys():
+        adj_list.append(adj1)
+        adj_list.append(adj2)
+    adj_list = list(set(adj_list))
+
+    MIN_THRESHOLD = 0.05
+    MAX_THRESHOLD = 0.15
```

Comment:
Add a brief comment for choosing these values. There should be detailed explanation in the doc but adding  a small comment with an example would help here.